### PR TITLE
[WIP] Fixes emailservice CrashLoop when Workload Identity is enabled

### LINKED ISSUE #255: emailservice crashlooping
Installed on Kubernetes via [mesh expansion example](https://github.com/GoogleCloudPlatform/istio-samples/tree/master/mesh-expansion-gce). All the other services are operating correctly, emailservice is crashlooping.

I set the cluster up without Workload Identity but added it later (enabled it on the cluster and bounced the nodepool to install the agent, [as per the instructions](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity))

```
{"timestamp": 1571526075.3596034, "severity": "INFO", "name": "emailservice-server", "message": "starting the email service in dummy mode."}
{"timestamp": 1571526075.3601294, "severity": "INFO", "name": "emailservice-server", "message": "Skipping Stackdriver Profiler Python agent initialization. Set environment variable ENABLE_PROFILER=1 to enable."}
INFO:emailservice-server:starting the email service in dummy mode.
INFO:emailservice-server:Skipping Stackdriver Profiler Python agent initialization. Set environment variable ENABLE_PROFILER=1 to enable.
INFO:emailservice-server:listening on port: 8080
{"timestamp": 1571526075.44175, "severity": "INFO", "name": "emailservice-server", "message": "listening on port: 8080"}
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526076.651824820","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526076.743373515","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526080.860304954","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526081.140930873","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526085.953478041","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526086.148386809","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
ERROR:grpc._server:Exception calling application: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 533, in __call__
    return _end_unary_response_blocking(state, call, False, None)
  File "/usr/local/lib/python3.7/site-packages/grpc/_channel.py", line 467, in _end_unary_response_blocking
    raise _Rendezvous(state, None, None, deadline)
grpc._channel._Rendezvous: <_Rendezvous of RPC that terminated with:
	status = StatusCode.UNAUTHENTICATED
	details = "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project."
	debug_error_string = "{"created":"@1571526090.846417037","description":"Error received from peer","file":"src/core/lib/surface/call.cc","file_line":1017,"grpc_message":"Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.","grpc_status":16}"
>

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/grpc/_server.py", line 389, in _call_behavior
    return behavior(argument, context), True
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/ext/grpc/server_interceptor.py", line 85, in new_behavior
    execution_context.get_opencensus_tracer().end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracer.py", line 117, in end_span
    self.tracer.end_span()
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/tracers/context_tracer.py", line 123, in end_span
    self.exporter.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 234, in export
    self.transport.export(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/transports/sync.py", line 23, in export
    self.exporter.emit(span_datas)
  File "/usr/local/lib/python3.7/site-packages/opencensus/trace/exporters/stackdriver_exporter.py", line 225, in emit
    self.client.batch_write_spans(project, stackdriver_spans)
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/client.py", line 100, in batch_write_spans
    name=name, spans=spans, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace/_gapic.py", line 89, in batch_write_spans
    name=name, spans=spans_pb_list, retry=retry, timeout=timeout
  File "/usr/local/lib/python3.7/site-packages/google/cloud/trace_v2/gapic/trace_service_client.py", line 256, in batch_write_spans
    request, retry=retry, timeout=timeout, metadata=metadata
  File "/usr/local/lib/python3.7/site-packages/google/api_core/gapic_v1/method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/google/api_core/grpc_helpers.py", line 59, in error_remapped_callable
    six.raise_from(exceptions.from_grpc_error(exc), exc)
  File "<string>", line 3, in raise_from
google.api_core.exceptions.Unauthenticated: 401 Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.

```

### PR DESCRIPTION
**Pull request by askmeegs**
Closes #255 

- [x]  updates emailservice init tracing logic to include an explicit `PROJECT_ID` for the `stackdriver_exporter`. When this was not specified, and workload identity was enabled on the cluster, this caused an authentication error. Adding the project ID makes it work, not entirely sure why, but I think it has to do with workload identity pods being decoupled from google service accounts - so you have to be explicit about the project. 
- [x]  I noticed that recommendationservice (the other Python service) had not fully enabled tracing with the `tracer_interceptor`, uncommented it, seems to be working fine. 
- [x] also tested changes on a Kops on GCE cluster

### CODE REVIEW

**Comment by ahmetb** [comment]
Thanks a lot for taking a stab at this, some questions:

- the behavior seems inconsistent (i.e. doesn't happen on other languages) it might be worth investigating why. (I suspect the metadata endpoint to get /project-id is somehow a different one than the others.)

- I wonder if it would've happened to recommendationservice as well –now that you've enabled tracing, and it's Python, too.

- where does `GCP_PROJECT_ID` env var comes from? It doesn't seem like we're setting it anywhere. What's the behavior of setting this to empty value (vs not setting at all)?

I'm trying to make sure (1) we go to the bottom of this, if there's a library bug in Stackdriver or OC (2) we don't introduce a new env var required to make it work on GCP –especially when other languages are working ok.

**Comment by askmeegs** [comment]
> Thanks a lot for taking a stab at this, some questions:
> 
> * the behavior seems inconsistent (i.e. doesn't happen on other languages) it might be worth investigating why. (I suspect the metadata endpoint to get /project-id is somehow a different one than the others.)

Good point. The example in the the opencensus/python docs has you [specify a project ID](https://github.com/census-instrumentation/opencensus-python/blob/f2231a56860e1adaf705d95a57437b3538858245/contrib/opencensus-ext-stackdriver/README.rst#trace). The init call to the metadata API, with the project ID, is [here](https://github.com/census-instrumentation/opencensus-python/blob/f4d8ec36cada8b385821a16795f27fe322694a1a/contrib/opencensus-ext-stackdriver/opencensus/ext/stackdriver/trace_exporter/__init__.py#L110). Still digging into what might be different about this. 

> * I wonder if it would've happened to recommendationservice as well –now that you've enabled tracing, and it's Python, too.
> * where does `GCP_PROJECT_ID` env var comes from? It doesn't seem like we're setting it anywhere. What's the behavior of setting this to empty value (vs not setting at all)?
> 

You're right, we aren't setting that. Inside the emailservice server container:

```
# echo $GCP_PROJECT_ID

```

And also, even though the service wasn't erroring, I realized that I wasn't seeing any traces in Stackdriver for this cluster. When I connected the default K8s namespace/svc account with a new Google Service Account (using [these instructions](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity#enable_workload_identity_on_a_new_cluster)), and I made that GSA a Cloud Trace Admin, traces started showing up: 

Screen Shot 2019-12-16 at 9 54 29 AM

I then reverted emailservice back to what's on master - project ID env var still not set, Workload identity still set up - and while I don't see errors, email service seems to be running without tracing: 

```
emailservice-775b96c487-6jbf8 server {"timestamp": 1576509028.4202971, "severity": "INFO", "name": "emailservice-server", "message": "A request to send order confirmation email to someone@example.com has been received."}
emailservice-775b96c487-6jbf8 server {"timestamp": 1576509037.8868206, "severity": "INFO", "name": "emailservice-server", "message": "A request to send order confirmation email to someone@example.com has been received."}
emailservice-775b96c487-6jbf8 server {"timestamp": 1576509046.6041756, "severity": "INFO", "name": "emailservice-server", "message": "A request to send order confirmation email to someone@example.com has been received."}
``` 

I re-add the init code where I set the project ID to the env var, and tracing logs come back:

```
emailservice-694687f95b-5c5k9 server [SpanData(name='Recv.hipstershop.EmailService.SendOrderConfirmation', context=SpanContext(trace_id=215639c02e16d3aafe57dd5231d74e26, span_id=4e0e8ddf14820203, trace_options=TraceOptions(enabled=True), tracestate=None), span_id='dc5c7865e03d40a1', parent_span_id='4e0e8ddf14820203', attributes={'component': 'grpc'}, start_time='2019-12-16T15:14:35.219322Z', end_time='2019-12-16T15:14:35.219727Z', child_span_count=0, stack_trace=None, time_events=[<opencensus.trace.time_event.TimeEvent object at 0x7f38bbaede10>, <opencensus.trace.time_event.TimeEvent object at 0x7f38bbaede50>], links=[], status=None, same_process_as_parent_span=None, span_kind=1)]
```

> I'm trying to make sure (1) we go to the bottom of this, if there's a library bug in Stackdriver or OC (2) we don't introduce a new env var required to make it work on GCP –especially when other languages are working ok.

So TLDR it looks like you both have to 1) set up workload identity for your namespace and 2) include the GCP_PROJECT_ID placeholder `""` on setup. 

I think this issue boils down to 2 things:
1. A discrepancy in how the python opencensus/stackdriver init catches authentication errors, vs. other languages. Not sure what it is yet. 
2. Pod identity not set up correctly on clusters where workload identity is enabled. It's on the user to do this setup. Maybe we add a line in the README pointing to the docs.

**Comment by ahmetb** [comment]
OK, since:

1. the behavior when locally is unchanged
2. when run on GKE, this fixes the crashloop, despite GCP_PROJECT_ID always being "" and it enables traces

we can merge.

File: kubernetes-manifests/emailservice.yaml [hunk 1]
```diff
@@ -34,6 +34,10 @@ spec:
         env:
         - name: PORT
           value: "8080"
+        - name: GCP_PROJECT_ID
+          value: "thinmint"
```
Comment (ahmetb) [inline kubernetes-manifests/emailservice.yaml]: undo before merge
